src/core/ddsi/src/q_entity.c

```diff
@@ -2288,7 +2290,13 @@ static void writer_add_local_connection (struct writer *wr, struct reader *rd)
             PGUID (wr->e.guid), PGUID (rd->e.guid));
   m->rd_guid = rd->e.guid;
   ddsrt_avl_insert_ipath (&wr_local_readers_treedef, &wr->local_readers, m, &path);
-  local_reader_ary_insert (&wr->rdary, rd);
+
+#ifdef DDS_HAS_SHM
+  if (!wr->has_iceoryx || !rd->has_iceoryx)
+    local_reader_ary_insert(&wr->rdary, rd);
+#else
+  local_reader_ary_insert(&wr->rdary, rd);
```

Comment:
I know that technically this #elseif is not needed if we exclude the branch from the #ifdef but I think this way intention and formatting is more clear.